src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -485,6 +509,23 @@ public PartitionIterator fetchPage(PageSize pageSize, long queryStartNanoTime)
                 return pager.fetchPageInternal(pageSize, executionController);
             }
         }
+
+        public static class OffsetPager extends Pager
```

Comment:
What is the purpose of this? It is a bit surprising to me that it doesn't really use the offset and it seems to return everything as a single page.

Reply:
I guess it is for disabling paging, is it true?

Reply:
Maybe consider naming it differently? It looks to me like all it does is concatenating all pages into one huge page.

Reply:
It indeed disables regular client-based paging, but it also replaces it with internal paging. 

That is, `fetchPage` will return a partition iterator that actually consumes all the rows, as if it was a big single page. However, the pager specified by the client's fetch size will still be internally used by the returned partition iterator. That way, the partition iterator won't read (and materialize!) all the data with a single read command. It rather would lazily consume results with as many read commands as needed.

This is because we don't need to keep the first `OFFSET` rows in memory since we are going to discard them. Thus, this should avoid the need to fetch the required `OFFSET+LIMIT` rows in one go, reducing memory usage.

Reply:
I have tried to add some JavaDoc about this.

Reply:
Ok, I have done some refactoring to shed some light on this. Rather than redefining the meaning of `Pager#fetchPage`, I have added a new `Pager#readAll` method that return an iterator on all the rows in the pager. That iterator uses the provided page size to lazily fetch the replicas as the returned iterator is consumed. 

That way, we can still internally use paging to return a single page to the client, but in a (hopefully) less confusing way.

I have also moved the internals from `SelectStatement.Pager` to `QueryPager`, so we can offload some responsibilities from the already overloaded `SelectStatement`.

Reply:
> Ok, I have done some refactoring to shed some light on this. Rather than redefining the meaning of Pager#fetchPage, I have added a new Pager#readAll method that return an iterator on all the rows in the pager

👍